### ISSUE TITLE
Re-write Dev Guide

### ISSUE THREAD

**Issue by brodyberg** [Microsoft/visualfsharp#64]
Current devguide is focused on simply getting to the point of being able to make changes. What is needed is additional content for the points beyond - the code organization, which things do what and why, what the performance considerations are for certain areas, suggested style for given regions and lastly we need something that describes the characteristics of great PRs which contain experiments vs. bug fixes vs. features etc.

**Comment by latkin** [comment]
There is an internal mail thread going which is very closely related to this.  I guess no reason not to migrate to a more open forum.

@dsyme plans to start putting together a true development guide for the compiler (with the kind of topics you mention), we are discussing possible options.

- Dedicated GitHub pages repo for the guide
- Part of wiki on this repo
- other?

And for any of those options, what is contribution model? Fully open to public editing (i.e. open wiki), or PR model?

**Comment by forki** [comment]
Personally I'm not a big fan of wikis and would prefer github pages approach with pull requests. But I'd recommend to create a bigger circle of trusted people who can accept these pull requests. It's much easier to accept/merge docs than compiler internals. It's also much easier to "repair"  wrong commits.

**Comment by brodyberg** [comment]
Wikis are over. Let's do the pages/PR model. 
+1 for adjusting the circle. Not "bigger is better" but rather "people with roles so no one gets burned out."

**Comment by jwood803** [comment]
If I may put my 2 cents in, I concur with what @forki has mentioned. GitHub pages seems to be better for documentation than wikis in my experience.

**Comment by dsyme** [comment]
I've started a guide at http://fsharp.github.io/2015/09/29/fsharp-compiler-guide.html.  

Currently I have placed it on the F# Core Engineering Group pages] (the F# Core Engineering Group is a technical working group of the F# Software Foundation but really encompasses any contributors to any core F# repositories).  It's possible the guide should be incorporated into the F# compiler repositories themselves though then publishing it becomes a bit more of a chore (currently publication is done automatically through github pages).
